django/contrib/auth/middleware.py

```diff
@@ -34,6 +39,68 @@ def process_request(self, request):
         request.auser = partial(auser, request)
 
 
+class LoginRequiredMiddleware:
+    """
+    Middleware that forces all views to require login by default.
+
+    Views that have login_not_required decorator or LoginNotRequiredMixin mixin
+    will be able to pass through without this validation. Otherwise, it will
+    redirect user to login or admin's login if accessing admin site.
+    """
+
+    sync_capable = True
+    async_capable = True
+
+    def __init__(self, get_response):
+        if get_response is None:
+            raise ValueError("get_response must be provided.")
+        self.get_response = get_response
+        self.async_mode = iscoroutinefunction(self.get_response)
+        if self.async_mode:
+            markcoroutinefunction(self)
+        super().__init__()
+
+    def __call__(self, request):
+        if not hasattr(request, "user"):
+            raise ImproperlyConfigured(
+                "The Django login required middleware requires authentication "
+                "middleware to be installed. Edit your MIDDLEWARE setting to "
+                "insert "
+                "'django.contrib.auth.middleware.AuthenticationMiddleware' "
+                "before "
+                "'django.contrib.auth.middleware.LoginRequiredMiddleware'."
+            )
+        if self.async_mode:
+            return self.__acall__(request)
+        self.user_is_authenticated = request.user.is_authenticated
+        return self.get_response(request)
+
+    async def __acall__(self, request):
+        self.user_is_authenticated = (await request.auser()).is_authenticated
```

Comment:
The current approach is insecure. Storing values on `self` allows those values to be shared across threads and async tasks. This leads to the possibility that two concurrent requests to a threaded/async server will mix up the “is authenticated” values and incorrectly show a private page.

Please change the middleware to inherit from `MiddlewareMixin`, drop the custom `__init__`, `__call__` and `__acall__`, and only implement `process_view`.

Reply:
Thanks for confirming! Reverted async stuff in https://github.com/django/django/pull/17792/commits/3dd63ed05bef5eabf016375af9c2be53a200f62d